Docs: Relabel Automatic Redirects as "Incoming links: Best practices and redirects" (Diátaxis)

benjaoming opened the pull request:
After reading this, I decided to expand the scope to deal more generally with the question of how to handle incoming links.

I think this would be the main theme that the reader is tackling when reading the explanation.

I added a lot of cross-references already. Potentially, more could be added.. but I got a feeling that the page would drown in cross-references at some point.

Other than that, I think this looks like a fine improvement. All the contents were already "explanation" so pretty easy to move.

Refs: #9746 

<!-- readthedocs-preview docs start -->
---
:books: Documentation previews :books:

- User's documentation (`docs`): https://docs--9896.org.readthedocs.build/en/9896/

<!-- readthedocs-preview docs end -->

<!-- readthedocs-preview dev start -->
- Developer's documentation (`dev`): https://dev--9896.org.readthedocs.build/en/9896/

<!-- readthedocs-preview dev end -->

humitos commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
+* Keep development versions hidden so people do not find them on search engines by mistake.
+* Avoid using a public ``latest`` for development versions and do not make your *default version* a development branch.
+  Publishing development branches can mean that users are reading instructions for unreleased software or draft documentation.
+* Try to get your understanding of ``latest`` and *default version* right from the beginning and ensure you don't change them later on.
```
This also seems to be good advice. I think I'm a little confused about this "Questionable practice" section 🤔

benjaoming replied:
I'll put the end of the section at the beginning :+1:

benjaoming replied:
I moved it to best practice :)

humitos commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
+   it can make the maintainer of the referring website aware that they need to update their link.
+   Users will be aware that the documentation still exist but has changed.
+
+   The default Read the Docs 404 page is designed to be helpful.
+   You can also design a custom 404 page, see :ref:`hosting:Custom Not Found (404) Pages`.
```
```suggestion
   The default Read the Docs 404 page is designed to be helpful,
   but you can also design your own matching the style of your site, see :ref:`hosting:Custom Not Found (404) Pages`.
```

humitos reviewed (commented):
I left a bunch of comments. The content is _good_; but I think there are some parts that can be clarified more to avoid confusions.

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
+
+* Use ``/page/path/to/page.html`` if you are linking to the page in the default version of the default language.
+* Use ``/en/latest/path/to/page.html`` if you want to be specific about the language.
+* If you move a page that likely has incoming references, :doc:`create a redirect rule </user-defined-redirects>`.
```
Maybe we should suggest to use Traffic Analytics to be informed before making decisions to (re)move content.

benjaoming replied:
I think we shouldn't write this "bombastically" - it should be possible to move a page without worrying too much. Later we make the point about 404s being useful.

humitos replied:
If we are adding a section about "Good recommendations", I'd say we should recommend to _always_ create a redirect if you change a URL. No matter if there were too many pageviews in the last month(s) or not. You never know where your URL is linked from 😄

ericholscher commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
 
 You can link to a specific page and have it redirect to your default version.
 This is done with the ``/page/`` URL prefix.
 You can reach this page by going to https://docs.readthedocs.io/page/automatic-redirects.html.
```
I think that just adds words without saying anything. The original seems clear?

humitos replied:
Maybe it's a personal thing. I usually find hard to parse "this" when there are many things involved since it depends on the context where it's written and can be ambiguos. In those cases, I tend to replace "this" with the exact thing I want to highlight in that sentence.

benjaoming replied:
I agree with @humitos that "this" can easily get confusing here. So I changed it. Managed to eliminate one word, so Hemingway would hopefully approve :)

ericholscher reviewed (commented):
Manuel did a great copy edit here, and I agree with most of the feedback. I'd also give some feedback at a high level which seems like this page is trying to mix reference & explanation and it doesn't work well. The section on `permalinks` feels really odd and different in tone than the rest of the article.

I think we should probably consider moving this back to a feature reference and an explanation. I also wouldn't frame the explanation as a question, since we don't do that anywhere else? Similarly, the best practices & "what not to do" seem odd for a explanation -- though I'm not 100% sure where they belong.

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
+Root URL redirect
+-----------------
 
 A link to the root of your documentation will redirect to the *default version*,
```
I like both suggestions! But they were targeting the same line so I edited your comment to have both suggestions "in one"

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
+
+   This only works for the root URL, not for internal pages.
+   It's designed to redirect people from ``http://pip.readthedocs.io/`` to the default version of your documentation,
+   since serving up a 404 here would be a pretty terrible user experience.
```
I think this is @ericholscher talking :) I liked it when I read it the first time, so I just left it.. but of course always safer to use neutral language. Let's figure this out some time in the style guide.. it's nice to have some kind of personal touch, but we have to know the _style_.

ericholscher replied:
Heh yea -- sounds like something I would write :)

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
+   since serving up a 404 here would be a pretty terrible user experience.
+
+.. note::
+   If the "develop" branch was designated as the default version,
+   then ``http://pip.readthedocs.io/`` would redirect to ``http://pip.readthedocs.io/en/develop``.
```
Agreed, gonna fix that.

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
+   So, for example, a link to an internal page like
+   ``http://pip.readthedocs.io/usage.html`` doesn't redirect to ``http://pip.readthedocs.io/en/latest/usage.html``.
+
+   The reasoning behind this is that RTD organizes the URLs for docs so that multiple translations and multiple versions of your docs can be organized logically and consistently for all projects that RTD hosts.
+   For the way that RTD views docs,
+   ``http://pip.readthedocs.io/en/latest/`` is the root directory for your default documentation in English, not ``http://pip.readthedocs.io/``.
+   Just like ``http://pip.readthedocs.io/en/develop/`` is the root for your development documentation in English.
```
Yes, this needs a clarification... let's make some `note::` or `tip::` about this. I'll come up with a suggestion.

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
+--------------------------
 
-The reasoning behind this is that RTD organizes the URLs for docs so that multiple translations and multiple versions of your docs can be organized logically and consistently for all projects that RTD hosts. For the way that RTD views docs, ``http://pip.readthedocs.io/en/latest/`` is the root directory for your default documentation in English, not ``http://pip.readthedocs.io/``. Just like ``http ...
+You might be familiar with permalinks from blogging.
+The idea is that a blog post receives a unique link as soon as it's published and that the link does not change afterward.
+Incoming sources can reference the blog post even though the blog changes structure or the post title changes.
```
I'm a bit reluctant to include it in our Glossary since we do not have _permalinks_ officially. I don't know if any documentation frameworks support this? I would love if they did, but I wanted to use the term so that people will understand that writing documentation and using a URL path is in some sense the same as creating a permalink.

humitos replied:
> I'm a bit reluctant to include it in our Glossary since we do not have permalinks officially 

Describing a term in the glossary does not mean that we officially support it. It helps us to expand the meaning of a word that in the same way through out the whole documentation, to give the user more context without extending the explanation of the word in all the pages.

In the glossary, we can explain things that we don't support at all; but we mention them in our documentation.

benjaoming replied:
Sounds good! It also aligns with an idea of being "friendly" to readers, i.e. make sure that they feel included when new terms are explained.

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
+Good practice ✅
+~~~~~~~~~~~~~~~~
+
+* Use ``/page/path/to/page.html`` if you are linking to the page in the default version of the default language.
+* Use ``/en/latest/path/to/page.html`` if you want to be specific about the language.
```
That will work well, yes! Good with a reference :sparkle:

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
+* Avoid using specific versions in links unless you really mean that users should see that exact version.
+  Versions get outdated.
+  You can use a :ref:`version warning <versions:Version warning>` to ensure the reader is aware.
+* Keep development versions hidden so people do not find them on search engines by mistake.
```
Haha, yes, good catch, I must have been trying to write "positive" in a section that should phrase things in a negative way :D

benjaoming replied:
Fixed!

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
+Shortlinks: rtfd.io
+-------------------
+
+Links to ``readthedocs.io`` are treated the same way as ``readthedocs.io``.
```
Not sure what happened here :scream: Will fix!

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
+and in the interest of handling "logical-looking" URLs (leaving out the lang_slug or version_slug shouldn't result in a 404),
+the following redirects are supported::
 
     /          -> /en/latest/
     /en/       -> /en/latest/
```
Oh! I just added "legacy", maybe it's beyond deprecated.

benjaoming replied:
I'd suggest to just remove the whole section then?

humitos replied:
IMO, removing the whole section for things that don't work anymore as they say it works is 👍🏼

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,56 +1,140 @@
+* Use ``/en/latest/path/to/page.html`` if you want to be specific about the language.
+* If you move a page that likely has incoming references, :doc:`create a redirect rule </user-defined-redirects>`.
+* Links from other Sphinx projects should use :doc:`intersphinx </guides/intersphinx>`.
+* Use minimal page slugs that do not attract renaming suggestions.
```
Yes, that one has a :term: - just need to use it :smile:

benjaoming replied:
Also fixed!

benjaoming commented:
> I'd also give some feedback at a high level which seems like this page is trying to mix reference & explanation and it doesn't work well.

After your comment, I changed a lot here, both the title and structure and intro.

I think the "reference" part of the article is really minimal, especially after we removed some incorrect old legacy section. The remaining reference isn't really a type of reference I'd expect anyone to visit a documentation page to look for. So ultimately, I think that the entire content is better shipped as an explanation with focus on how to handle external links. Through my own "fresh eyes" reading, I think this goal is near after the latest changes :+1:

ericholscher commented:
I will give it a read over tomorrow. I ran out of energy today 😴

ericholscher commented:
Talked about this on our call today. I think we have good next steps:

* Combine User-defined & Automatic redirects into a `Redirects` reference
* Keep the explanation at a high-level around best practices and permalink explainer
*

benjaoming commented:
Just noticed that we have this page: https://docs.readthedocs.io/en/stable/guides/deprecating-content.html

benjaoming commented on docs/user/guides/redirects.rst:
```diff
@@ -0,0 +1,107 @@
+  With this option,
+  you can specify a page in your documentation to redirect elsewhere.
+  The rule triggers no matter the version of your documentation that the user is visiting.
+  This rule can also redirect to another website.
```
This is from the original reference. However, the preview function doesn't reflect this. Either that's a bug or the information in the reference was incorrect. Will need to check this...

benjaoming replied:
It checks out, so it's a bug in the UI that I already reported.

benjaoming commented:
@ericholscher what you mentioned in https://github.com/readthedocs/readthedocs.org/pull/9896#issuecomment-1398839627 from our earlier call this week is implemented here and ready for review :+1:

ericholscher commented on docs/user/user-defined-redirects.rst:
```diff
@@ -1,39 +1,116 @@
+Built-in redirects
+------------------
+
+This article explains the redirects that are automatically active for all Read the Docs projects and how they are useful.
+Automatic redirects on Read the Docs are supported long-term.
+
+Built-in redirects are especially useful for creating and sharing **incoming links**
+which is discussed indepth in :doc:`/automatic-redirects`.
```
```suggestion
This section explains the redirects that are automatically active for all projects and how they are useful.
Built-in redirects are especially useful for creating and sharing incoming links,
which is discussed indepth in :doc:`/automatic-redirects`.
```

ericholscher commented on docs/user/user-defined-redirects.rst:
```diff
@@ -1,39 +1,116 @@
+
+For example::
+
+    docs.readthedocs.io -> docs.readthedocs.io/en/latest/
+    www.pip-installer.org -> www.pip-installer.org/en/latest/
```
```suggestion
    docs.readthedocs.io -> docs.readthedocs.io/en/stable/
```

ericholscher commented on docs/user/user-defined-redirects.rst:
```diff
@@ -1,39 +1,116 @@
+
+.. warning::
+
+   This only works for the root URL, not for internal pages.
```
What are internal pages?

benjaoming replied:
I don't know + this whole warning box is weird. Look at the warning text below.

Gonna delete it...

benjaoming replied:
I think we should combine the two warnings to a single `note::` that says

> You cannot use the root redirect to reference specific pages. `/` redirects to the default version, but `/some/page.html` will not redirect to `/en/latest/some/page.html`. Use the `/page/` redirect rule for this purpose.

benjaoming replied:
Made it a warning box.

ericholscher commented on docs/user/user-defined-redirects.rst:
```diff
@@ -1,39 +1,116 @@
+   It's designed to redirect people from `/` to the default version of your documentation,
+   since serving up a 404 here would be bad user experience.
+
+.. note::
+   If the "develop" branch was designated as the default version,
+   then ``http://pip.readthedocs.io/`` would redirect to ``http://pip.readthedocs.io/en/develop``.
+   But, it's not a universal redirecting solution.
+   So, for example, a link to an internal page like
+   ``http://pip.readthedocs.io/usage.html`` doesn't redirect to ``http://pip.readthedocs.io/en/latest/usage.html``.
+
+   The reasoning behind this is that RTD organizes the URLs for docs so that multiple translations and multiple versions of your docs can be organized logically and consistently for all projects that RTD hosts.
+   For the way that RTD views docs,
+   ``http://pip.readthedocs.io/en/latest/`` is the root directory for your default documentation in English, not ``http://pip.readthedocs.io/``.
+   Just like ``http://pip.readthedocs.io/en/develop/`` is the root for your development documentation in English.
+
```
```suggestion
```

Seems this both of these notes say the same thing? I don't like using examples other people control, so we should just remove it.

ericholscher commented on docs/user/user-defined-redirects.rst:
```diff
@@ -1,39 +1,116 @@
+   ``http://pip.readthedocs.io/en/latest/`` is the root directory for your default documentation in English, not ``http://pip.readthedocs.io/``.
+   Just like ``http://pip.readthedocs.io/en/develop/`` is the root for your development documentation in English.
+
+Among all the multiple versions of docs,
+you can choose which is the "default" version for Read the Docs to display,
+which usually corresponds to the git branch of the most recent official release from your project.
+
+Shortlink with ``https://*.rtfd.io``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Links to ``rtfd.io`` are treated the same way as ``readthedocs.io``.
+They redirect the root URL to the default version of the project.
+They are intended to be easy and short for people to type.
+
+You can reach these docs at https://docs.rtfd.io.
```
```suggestion
You can choose which is the "default" version for Read the Docs to display.
This usually corresponds to version of the most recent official release from your project.

Shortlink with ``https://<slug>.rtfd.io``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Links to ``rtfd.io`` are treated the same way as ``readthedocs.io``.
They are intended to be easy and short for people to type.

You can reach these docs at https://docs.rtfd.io.
```

benjaoming replied:
Thanks! Modified your suggestion a tiny bit :+1:

ericholscher reviewed (commented):
Lots of small nitpicks, but overall this is good.

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -82,63 +51,40 @@ you most likely want users clicking on incoming links to see the latest version
-* Use ``/page/path/to/page.html`` if you are linking to the page in the default version of the default language.
-* Use ``/en/latest/path/to/page.html`` if you want to be specific about the language.
-* If you move a page that likely has incoming references, :doc:`create a redirect rule </user-defined-redirects>`.
+* Use `page redirects <user-defined-redirects:Page redirects>`_ if you are linking to the page in the default version of the default language.
```
:100:

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -82,63 +51,40 @@ you most likely want users clicking on incoming links to see the latest version
-* If you move a page that likely has incoming references, :doc:`create a redirect rule </user-defined-redirects>`.
+* Use `page redirects <user-defined-redirects:Page redirects>`_ if you are linking to the page in the default version of the default language.
+* If you move a page that likely has incoming references, :doc:`create a custom redirect rule </guides/redirects>`.
 * Links from other Sphinx projects should use :doc:`intersphinx </guides/intersphinx>`.
```
good question... I guess from an author's POV it's a bit of a hard mission to check if incoming sources are using intersphinx.

Seems good to also mention "thou shalt cross-reference other Sphinx projects as thou wanteth other documentation projects to cross-reference your Sphinx project"

Must be better to not live in a glasshouse... also not throw stones actually

benjaoming replied:
What I mean is that I think it's cool to also mention tips about how to reference other Sphinx projects in a good way :+1:

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -82,63 +51,40 @@ you most likely want users clicking on incoming links to see the latest version
   keep original file names rather than going for low-impact URL renaming.
   Renaming an article's title is great for the reader and great for SEO,
   but this does not have to involve the URL.
+* Try to get your understanding of ``latest`` and *default version* right from the beginning and ensure you don't change them later on.
```
`latest` refers to a URL path, but it's not so clear

Will try with this:

```suggestion
* Establish your understanding of the *latest* and *default version* of your documentation at the beginning. Changing their meaning is very disruptive to incoming links.
```

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -82,63 +51,40 @@ you most likely want users clicking on incoming links to see the latest version
   but this does not have to involve the URL.
+* Try to get your understanding of ``latest`` and *default version* right from the beginning and ensure you don't change them later on.
+* Keep development versions hidden so people do not find them on search engines by mistake.
+  Publishing documentation at the wrong URL can be costly if the URL starts circulating.
```
```suggestion
  This is the best way to ensure that nobody links to URLs that are intended for development purposes.
```

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -82,63 +51,40 @@ you most likely want users clicking on incoming links to see the latest version
 ~~~~~~~~~~~~~~~~~~~~~~~~
 
-* Avoid using specific versions in links unless you really mean that users should see that exact version.
+* Avoid using specific versions in links unless you really intend users to see that exact version.
```
I think the sentence is a bit more cryptic now, but it's shorter :+1:

ericholscher replied:
Not sure I'd say it's more cryptic, and not sure if this response is sarcasm?

benjaoming replied:
No, it wasn't. When text is trimmed out, *some* meaning can be lost. In this case, it's about clarifying and spelling out *unless you really intend users to see that exact version*. Because the context is a list of "Questionable practice 🟡", then the idea of spelling out why the practice can be questionable is a bit more in the spotlight :)

When shortening to *unless users need that exact version*, there's the risk that someone doesn't get the meaning."Yes, they should see that version". The overly long sentence was supposed to capture the afterthought.

Maybe not successful, I get that :+1:

benjaoming commented on docs/user/glossary.rst:
```diff
@@ -7,10 +7,25 @@ Glossary
       `Main page <https://readthedocs.org/dashboard>`_ where you can see all your projects with their build status
       and import a new project.
 
+   default version
```
This is a mistake. Will add references.

benjaoming commented on docs/user/glossary.rst:
```diff
@@ -7,10 +7,25 @@ Glossary
       and import a new project.
 
+   default version
+      A documentation has a *default version*, usually understood to be the latest stable version of a project.
+      Most documentation projects will contain both older versions of documentation and the latest development version.
+      The *default version* can be fully configured to point to a specific Git branch or any other released version.
+      In the most common scenario,
+      the default version points to the ``main`` branch of a Git repository since it usually points to the latest stable release.
```
That's a much clearer way to explain it :+1: Also nice with the URL-perspective since it was intended to be included in pages about URLs and links :)

benjaoming commented on docs/user/automatic-redirects.rst:
```diff
@@ -1,72 +1,41 @@
---------------------------
-
-You might be familiar with permalinks from blogging.
+You might be familiar with :term:`permalinks <permalink>` from blogging.
```
it's a very good article that I used as inspiration for the glossary entry :)

but yeah, I'll trim that entry since it gives the impression that it's a term that RTD has a special definition of.

benjaoming commented:
@ericholscher PR is ready for another review :eyes:

ericholscher reviewed (approved):
Changes look good 👍
